metric_learn/constraints.py

```diff
@@ -33,6 +35,56 @@ def positive_negative_pairs(self, num_constraints, same_length=False,
       return a[:n], b[:n], c[:n], d[:n]
     return a, b, c, d
 
+  def generate_knntriplets(self, X, k_genuine, k_impostor):
+
+    labels = np.unique(self.partial_labels)
+    L = len(labels)
+    len_input = np.size(self.partial_labels, 0)
+    triplets = np.empty((len_input*k_genuine*k_impostor, 3), dtype=np.intp)
+
+    start = 0
+    finish = 0
+    neigh = NearestNeighbors()
+
+    for i in range(L):
+
+        # generate mask for current label
+        gen_mask = self.partial_labels == labels[i]
+        gen_indx = np.where(gen_mask)
+
+        # get k_genuine genuine neighbours
+        neigh.fit(X=X[gen_indx])
+        gen_neigh = np.take(gen_indx, neigh.kneighbors(n_neighbors=k_genuine,
+                            return_distance=False))
+
+        # generate mask for impostors of current label
+        imp_indx = np.where(np.invert(gen_mask))
+
+        # get k_impostor impostor neighbours
+        neigh.fit(X=X[imp_indx])
+        imp_neigh = np.take(imp_indx, neigh.kneighbors(
+                            n_neighbors=k_impostor,
+                            X=X[gen_mask],
+                            return_distance=False))
+
+        # lenght = len_label*k_genuine*k_impostor
+        finish += np.sum(gen_mask)*k_genuine*k_impostor
+
+        triplets[start:finish, :] = self._comb(gen_indx, gen_neigh, imp_neigh,
+                                               k_genuine, k_impostor)
+        start = finish
+
+        # TODO: deal with too litle elements for k neighbors to be yielded
```

Comment:
yes, definitely need this, and some tests

Reply:
How should this be handled? Should there be a k_vector = np.minimum(k, labels_count) like it is handled on the LDA basis construction? or should an error be raised for the user to select a lower number for k?

I am more inclined to the second option.

Reply:
I decided to implement this similarly as it is done with the LDA basis construction, by creating a k-vector that adjusts k to the classes that need a lower value. Also, an informative warning is thrown for every class with this issue in order to allow the user to know this is the underlying behaviour. Let me know if you think this a good way to go, or if there is a better alternative!